### ISSUE TITLE
problem running basic optimization, require retain_grad=true

### ISSUE THREAD

**Issue by emauryg** [mlverse/torch#599]
I'm trying to implement the following optimization function, but the inner loop does not even go through the first backward call. 

```
transform_E <- function(mod_ts,Y,K, tau=50,lr=0.001, max_iter=10000){
  converged = FALSE
  old_loss = 1e10
  it = 0
  K = K
  D = Y$size(dim=-1)

  run_E = nn_module(
    classname= "run_E",
    initialize = function(E0){
      self$E_ = nn_parameter(E0)
    },
    forward = function(Y,T_tensor, tau, anno_dims){
      exp_E = torch_exp(self$E_)
      Chat = T_tensor$matmul(exp_E)
      D = ncol(exp_E)
      L = (-Y*torch_log(Chat)+ (Y+tau)*torch_log(Chat+tau))$sum()/D
      return(L)
    }
  )
  T_tensor = calculate_T_ts(mod_ts)$clone()
  E0 = torch_randn(c(K, D), device=device)
  tmp_E = run_E(E0)
  optimizer = optim_sgd(tmp_E$parameters,lr=lr)
  start = Sys.time()
  while(converged==FALSE && it<= max_iter){
    it = it + 1
    if(it %% 100 == 0){
      message(paste("Current iteration:", it,", loss =", round(old_loss,2)))
    }
    optimizer$zero_grad()
    new_loss = tmp_E(Y,T_tensor, tau=tau, anno_dims=anno_dims)
    new_loss$backward()
    optimizer$step()
    if (it %% 100 == 0){
      converged = stop_ts(old_loss, new_loss$item())
      old_loss = new_loss$item()
    }
  }

  E = torch_exp(tmp_E$parameters$E_$detach())
  message(paste("It took:", difftime(Sys.time(),start, unit="mins",
    " mins to converge.")))
  return(E)

}
```

I get the following error:

Error in (function (self, inputs, gradient, retain_graph, create_graph) : Trying to backward through the graph a second time, but the saved intermediate results have already been freed. Specify retain_graph=True when calling .backward() or autograd.grad() the first time.
Traceback:

1. transform_E(mod_ts = mod_ts, Y = count_matrix, K = K, tau = 50)
2. new_loss$backward()
3. private$`__backward`(gradient = gradient, inputs = inputs, retain_graph = keep_graph, 
 .     create_graph = create_graph)
4. call_c_function(fun_name = "_backward", args = args, expected_types = expected_types, 
 .     nd_args = nd_args, return_types = return_types, fun_type = "method")
5. do_call(f, args)
6. do.call(fun, args)
7. (function (self, inputs, gradient, retain_graph, create_graph) 
 . {
 .     invisible(.Call("_torch_cpp_torch_method__backward_self_Tensor_inputs_TensorList", 
 .         PACKAGE = "torchpkg", self, inputs, gradient, retain_graph, 
 .         create_graph))
 . })(self = <pointer: 0x5bf3b410>, inputs = NULL, gradient = list(), 
 .     retain_graph = FALSE, create_graph = FALSE)

which is weird since I did not implement anything novel compared to other optimization routines I've implemented. 
It is unclear to me why it would say it's doing a second loop through the graph if this is the first iteration. 

Thank you

**Comment by emauryg** [comment]
update: issue resolved. Turns out one of the tensors was an output from another optimization routine, but it was not detached.
